impl/src/main/java/org/ehcache/impl/internal/store/disk/OffHeapDiskStore.java

```diff
@@ -488,12 +488,12 @@ public void initAuthoritativeTier(AuthoritativeTier<?, ?> resource) {
    */
   @SuppressWarnings("unchecked")
   public static <T> PersistentPortability<T> persistent(final Portability<T> normal) {
-    final Class<?> normalKlazz = normal.getClass();
-    Class<?>[] delegateInterfaces = normalKlazz.getInterfaces();
-    Class<?>[] proxyInterfaces = Arrays.copyOf(delegateInterfaces, delegateInterfaces.length + 1);
-    proxyInterfaces[delegateInterfaces.length] = PersistentPortability.class;
-
-    return (PersistentPortability<T>) Proxy.newProxyInstance(normal.getClass().getClassLoader(), proxyInterfaces, (o, method, os) -> {
+    List<Class<?>> proxyInterfaces = new ArrayList<>();
+    proxyInterfaces.add(PersistentPortability.class);
+    for (Class<?> klazz = normal.getClass(); klazz != null; klazz = klazz.getSuperclass()) {
+      proxyInterfaces.addAll(asList(klazz.getInterfaces()));
```

Comment:
This may be OK in this circumstance, but there's a restriction on what can be passed to `Proxy.newInstance` -- "No two elements in the interfaces array may refer to identical Class objects".  Doing so will result in an `IllegalArgumentException`.  Is there any chance this restriction will be violated in this code as a surprise?

Reply:
Changed to a `Set<Class<?>` should be safe now.

Reply:
There is an order dependency as well so you should really use a `LinkedHashSet` to maintain inheritance order.

Reply:
How does this look?